### PR TITLE
Finished second milestone

### PR DESCRIPTION
**Pull request by MkrtichSargsyan**
In this milestone, I created different users who can send invitations , show their friends list, and also reject invitations

### CODE REVIEW

File: app/views/users/_friends_table.html.erb [hunk 1]
```diff
@@ -0,0 +1,32 @@
+ <section class='bg-light d-flex'>
+    <article class='col-6'>
+      <h2>Friends List</h2>
+      <ol>
+        <% @user.friends and @user.friends.each do |friend| %>
+        <div class='mb-2'>
+          <li><%= friend.name %></li>
+        </div>
+        <% end %>
+        <%= "<b>No friends yet</b>".html_safe if @user.friends.empty? %>
```
Comment (lekegitrepo) [inline app/views/users/_friends_table.html.erb]: - **_There should be NO logic in the view_**

File: app/views/users/_friends_table.html.erb [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+      </ol>
+    </article>
+
+    <article class='col-6'>
+      <h2>Pending Requests</h2>
+      <ol>
+        <% @user.friend_requests and @user.friend_requests.each do |req| %>
+        <div class='d-flex align-items-center mb-2'>
+          <li><%= req.name %></li>
+          <% if @user == current_user %>
```
Comment (lekegitrepo) [inline app/views/users/_friends_table.html.erb]: - **_There should be NO logic in the view_**

File: app/views/users/_friends_table.html.erb [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+          
+            <%= button_to 'Accept', accept_user_path, :method => :get, params: { data: req.id },class:'p-1 ml-2' %>
+            <%= button_to 'Decline', decline_user_path, :method => :get, params: { data: req.id },class:'p-1 ml-2 bg-danger' %>
+          <% end %>
+          
+        </div>
+        <% end %>
+        <%= "<b>No friend requests</b>".html_safe if @user.friend_requests.empty? %>
```
Comment (lekegitrepo) [inline app/views/users/_friends_table.html.erb]: - **_There should be NO logic in the view_**

File: app/views/users/_user.html.erb [hunk 4]
```diff
@@ -1,6 +1,12 @@
...
\ No newline at end of file
+  <div class="profile-link ">
+    <span class='black-font'>
+      Name: 
+    </span>
+    <%= link_to user.name, user_path(user), class: 'mr-3 text-primary' %>
+
+     <% if !user.friends.include?(current_user) && user != current_user  && !user.pending? %>
+      <%= button_to 'send invitation', create_friendship_user_path(user.id), :method => :get %>
+    <% end %>
```
Comment (lekegitrepo) [inline app/views/users/_user.html.erb]: - **_There should be NO logic in the view_**

File: app/views/users/show.html.erb [hunk 5]
```diff
@@ -1,9 +1,17 @@
 <section class="user-section">
-  <h2><%= "Name: #{@user.name}" %></h2>
+  <div>
+    <h2><%= "Name: #{@user.name}" %></h2>
+    <% if !@user.friends.include?(current_user) && @user != current_user  && !@user.pending? %>
+      <%= button_to 'send invitation', create_friendship_user_path, :method => :get %>
+    <% end %>
+    <%= "<b>Pending request</b>".html_safe if @user.inverse_friendships.find{|f| f.user_id == current_user.id and !f.confirmed} %>
+    <%= "<b>Already friend</b>".html_safe if @user.friends.include?(current_user) %>
```
Comment (lekegitrepo) [inline app/views/users/show.html.erb]: - **_There should be NO logic in the view_**

**Review by lekegitrepo — CHANGES_REQUESTED** [review]
## Status: Changes Required! :recycle:
### Hi @MkrtichSargsyan :clap: :clap:
#### You have done a good job :+1: The associations between active records are well implemented. However, some requirements are missing/not implemented correctly. Check my comments below :point_down:

#### App Behaviour:
- _User should be able to see friendship requests ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/pending_request.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/)). Currently, the user does not know if they have a request(nothing indicate if a user has friend request)._
- _User has to accept a friendship request to start a friendship ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/)). There is no button to accept a friend request._

- _User can reject a friendship request ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/)). There is no button to reject a friend request._
- _There is a button to invite a user to friendship on the users' list page and on the single-user page ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/user_can_see_invite_buttton.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/)). The button appears when I start the app and the moment I sent a first friend request, the button did not appear again both in the users' list page and in the single-user page._
- _Posts timeline should include posts written by logged-in user and their friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/)). The posts are visible to any logged-in user right now either the author of the post is a friend of the current user or not._

#### Tests:
- Add more tests that cover the remaining models, controller, and(if possible you can add) integration tests to cover the typical user interaction with the application such as sign in, sign up, friend request, reject the request and accept the request.

#### Logic in the view:
- _There should be NO application logic in the views. Everything connected with data processing should be encapsulated in models and used in controllers._
- _View-related logic can be encapsulated in View helpers (app/helpers/\*_helper.rb)_
- _There should be absolutely NO logic in views. >Views are only for displaying information that became available to it from its corresponding controller not generating it. **Check these resources**_
* [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code)
* [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)
* [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright

#### _Reviewed by Temitayo Adeleke_

File: app/views/layouts/_navigation.html.erb [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+      <%= menu_link_to 'All users', users_path %>
+    </div>
+  </div>
+  <div class='d-flex align-items-center'>
+    <% if current_user %>
+      <p class='mr-3 mb-0'>Logged in as [<%= current_user.name %>]</p>
+      <%= link_to 'Sign out', destroy_user_session_path, method: :delete %>
+    <% else %>
+      <%= link_to 'Sign in', user_session_path %>
+    <% end %>
```
Comment (acushlakoncept) [inline app/views/layouts/_navigation.html.erb]: There should be NO logic in the view

File: app/views/layouts/application.html.erb [hunk 7]
```diff
@@ -8,22 +8,7 @@
...
-        </div>
-        <div class="menu-item">
-          <%= menu_link_to 'All users', users_path %>
-        </div>
-      </div>
-      <% if current_user %>
-        <%= link_to 'Sign out', destroy_user_session_path, method: :delete %>
-      <% else %>
-        <%= link_to 'Sign in', user_session_path %>
-      <% end %>
```
Comment (acushlakoncept) [inline app/views/layouts/application.html.erb:25]: There should be NO logic in the view

File: app/views/users/_friends_table.html.erb (hunk 1 above)
Comment (acushlakoncept) [inline app/views/users/_friends_table.html.erb]: There should be NO logic in the view

File: app/views/users/_user.html.erb [hunk 8]
```diff
@@ -1,6 +1,12 @@
...
\ No newline at end of file
+  <div class="profile-link ">
+    <span class='black-font'>
+      Name: 
+    </span>
+    <%= link_to user.name, user_path(user), class: 'mr-3 text-primary' %>
+
+     <% if !user.friends.include?(current_user) && user != current_user && !user.pending?(current_user)%>
+      <%= button_to 'send invitation', create_friendship_user_path(user.id), :method => :get %>
+    <% end %>
```
Comment (acushlakoncept) [inline app/views/users/_user.html.erb]: There should be NO logic in the view

File: app/views/users/show.html.erb [hunk 9]
```diff
@@ -1,9 +1,17 @@
 <section class="user-section">
-  <h2><%= "Name: #{@user.name}" %></h2>
+  <div>
+    <h2><%= "Name: #{@user.name}" %></h2>
+    <% if !@user.friends.include?(current_user) && @user != current_user  && !@user.pending?(current_user) %>
+      <%= button_to 'send invitation', create_friendship_user_path, :method => :get %>
+    <% end %>
+    <%= "<b>Pending request</b>".html_safe if @user.inverse_friendships.find{|f| f.user_id == current_user.id and !f.confirmed} %>
+    <%= "<b>Already friend</b>".html_safe if @user.friends.include?(current_user) %>
```
Comment (acushlakoncept) [inline app/views/users/show.html.erb]: There should be NO logic in the view

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

### Great job on your progress so far 👍 

Looks like you left out a couple of changes requested by the previous TSE, please, lets kindly address those:

- [ ] There should be absolutely NO logic in views. Views are only for displaying information that became available to it from its corresponding controller not generating it.  
See affected files below:
> * `layouts/_navigation.html.erb`
> * `layouts/application.html.erb`
> * `users/_friends_table.html.erb`
> * `users/_user.html.erb`
> * `users/show.html.erb`

**The Resources below may come in handy**
* [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code)
* [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)
* [Rails View Helpers Best Practices](https://medium.com/@acushla4real/rails-view-helpers-best-practices-guide-d7d7b6db3db4)
* [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright

- [ ] Add more tests coverage, you currently have tested just for the user model, kindly add for the remaining models, controller,  also include integration tests.
- [ ] It will be nice to also have the **Accept or Reject** request button on the users list page and on the single-user page

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla), if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

File: app/views/users/show.html.erb [hunk 10]
```diff
@@ -1,9 +1,16 @@
 <section class="user-section">
-  <h2><%= "Name: #{@user.name}" %></h2>
+  <div>
+    <h2><%= "Name: #{@user.name}" %></h2>
+   
+    <%= show_send_invitation_button(@user) %>
+    <%= "<b>Pending request</b>".html_safe if @user.inverse_friendships.find{|f| f.user_id == current_user.id and !f.confirmed} %>
+    <%= "<b>Already friend</b>".html_safe if @user.friends.include?(current_user) %>
```
Comment (acushlakoncept) [inline app/views/users/show.html.erb:7]: - [x] There should be NO application logic in the views nor helpers (no Post.new, Comment.create, current_user.friends.find_by(...)...). Everything connected with data processing should be encapsulated in models and used in controllers.

File: app/views/layouts/application.html.erb [hunk 11]
```diff
@@ -8,22 +8,7 @@
...
-    </nav>
+    <%= render 'layouts/navigation' %>
     <% if notice.present? %>
       <div class="notice">
```
Comment (acushlakoncept) [inline app/views/layouts/application.html.erb]: There should be NO logic in the view

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

### Great job! 👏  Both of you have done well on this project.

However, there are still some changes that are necessary for this project to be approved.
Let us quickly add some finishing touches and request another review.

- [ ] It will be nice to also have the Accept or Reject request button on the users' list page and on the single-user page
acc

- [x] I noticed you still have logic in your views, see the comments below and address them accordingly.

Having trouble? **The Resources below may come in handy**
* [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code)
* [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)
* [Rails View Helpers Best Practices](https://medium.com/@acushla4real/rails-view-helpers-best-practices-guide-d7d7b6db3db4)
* [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla), if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

**Review by SpaYco — APPROVED** [review]
# Project Approved

thank you for making the required changes, your project is now approved!

**_please be more professional with your commit messages, I understand having many TSEs asking for changes can be frustrating, but we shouldn't let frustration control our actions_**

### please don't forget to merge to master, good luck with the next projects!


Best, 
Aziz Mejri 
_contact me on [twitter](https://twitter.com/iSpaYco) or slack (@SpaYco) if you need any help_

**Pull request merged**